### PR TITLE
OPUSVIER-4213: Refactoring of improved BibTeX importer

### PR DESCRIPTION
**Pull request by saschaszott**
@j3nsch ich habe versucht alle deine Anmerkungen aus den beiden langen Feedback-E-Mails zu berücksichtigen.

Vorab kannst Du gern schon einen Blick auf das Resultat werfen.

Folgende Sachen sind noch in Bearbeitung:

- [x] Klärung `annote`
- [ ] Überarbeitung `README.md` nach Refactoring
- [ ] CLI-Command überarbeiten
- [ ] Integration in Administrationsbereich

Vielleicht hast Du aber dennoch Lust/Zeit schon jetzt zu schauen, ob alles in die richtige Richtung gelaufen ist.

### CODE REVIEW

**Comment by saschaszott** [comment]
> Den Build Break kann ich lokal mit dem Kommando
> 
> ```shell
> phpunit --log-junit build/phpunit.xml --colors=always --coverage-html build/coverage/ --coverage-clover build/coverage/clover.xml",
> ```
> 
> nicht reproduzieren.
> 
> @j3nsch , laufen die Tests lokal auf deiner Entwicklungsmaschine durch oder schlagen die dort auch fehl?

Ich vermute, dass es am sehr alten Pandoc liegt, das beim Hochfahren des Testcontainers installiert wird. Bei mir lokal auf einem **Ubuntu 20.10** ist Version 2.9.2.1 vom März 2020 (https://github.com/jgm/pandoc/releases/tag/2.9.2.1) installiert - auf dem Testcontainer allerdings noch die ziemlich alte Version 1.16 aus dem Januar 2016. Mit Pandoc 2 gab es viele Änderungen und _Breaking Changes_. Für den BibTeX-Import wird Pandoc 2 benötigt. Ich werde dazu einen Hinweis in die `README.md` aufnehmen. Wie kommen wir mit dem Testcontainer bei Github Actions weiter?

Ausgabe von meinem Entwicklungssystem:

```bash
$ pandoc  -v
pandoc 2.9.2.1
Compiled with pandoc-types 1.20, texmath 0.12.0.2, skylighting 0.8.5
Default user data directory: /home/sascha/.local/share/pandoc or /home/sascha/.pandoc
Copyright (C) 2006-2020 John MacFarlane
Web:  https://pandoc.org
This is free software; see the source for copying conditions.
There is no warranty, not even for merchantability or fitness
for a particular purpose.
```

Auch das ist nicht die aktuellste Version (neueste Pandoc-Version hat die Nummer 2.13).

**Comment by j3nsch** [comment]
Ja, bei meinem Ubuntu 16 schlagen die gleichen Tests fehl. Es wäre gut einen Hinweis in die README zu schreiben, am Besten mit Link zu PanDoc, damit man bei Bedarf manuell eine aktuellere Version installieren kann. Es sollte auch gesagt werden was genau nicht funktioniert mit dem alten PanDoc. Ist der BibTeX-Import trotzdem nutzbar?

Für die Tests auf GitHub werde ich die Testumgebung auf Ubuntu 20 umstellen.

**Comment by j3nsch** [comment]
Stell mal bitte in Deinem Branch die Workflow-Konfiguration auf Ubuntu 20 um und nimm dann auch gleich den Wechsel der PHP-Version aus `.github/workflows/php.yml` rauss.

**Comment by saschaszott** [comment]
> Ja, bei meinem Ubuntu 16 schlagen die gleichen Tests fehl. Es wäre gut einen Hinweis in die README zu schreiben, am Besten mit Link zu PanDoc, damit man bei Bedarf manuell eine aktuellere Version installieren kann. Es sollte auch gesagt werden was genau nicht funktioniert mit dem alten PanDoc. Ist der BibTeX-Import trotzdem nutzbar?
> 
> Für die Tests auf GitHub werde ich die Testumgebung auf Ubuntu 20 umstellen.

Die Abhängigkeit auf Pandoc 2.x habe ich in die `README.md` aufgenommen (ist aber noch nicht gepusht). Mit einem aktuellen OS sollte auch Pandoc 2 via Paketverwaltung installiert werden ohne Zusatzaufwand. Pandoc 1.x wurde seit mehr als 4 Jahren nicht mehr angefasst. Ich habe die Ursache der fehlgeschlagenen Tests jetzt nicht im Detail analysiert - aber offenbar verhält sich die Bibliothek bei bestimmten _Edge Cases_ anders. Wollen wir hier noch viel Energie reinstecken oder setzen wir einfach Pandoc 2 voraus. Der BibTeX-Import mit Pandoc 1 wird vermutlich in den meisten Fällen auch funktionieren - entwickelt und getestet habe ich es aber ausschließlich mit der o.g. Version (2.9.2.1).

**Comment by j3nsch** [comment]
Es war Quatsch von mir zu sagen, dass die PHP Umstellung weg kann. Wir verwenden ja auch im Hosting erst einmal noch eine alte PHP Version weiter. Also die PHP Version muss < 7.2 bleiben. 

Dieses konkrete Problem könnte vermutlich auch gelöst werden indem auf den Coding Style von Laminas umgestiegen wird, also nicht mehr ZF1 verwendet wird. Das BibTeX-Package hängt ja ansonsten nicht von Zend oder Laminas ab.

**Comment by saschaszott** [comment]
Leider unterstützt Ubuntu 20.04 nicht direkt PHP 7.1 (bzw. 7.0): https://github.com/marketplace/actions/setup-php-action#github-hosted-runners

Ich habe jetzt die Ubuntu Version auf 18.04 gesetzt. Damit kann ich dann zwar PHP 7.1 installieren; allerdings wird auch hier wieder die stark veraltete Pandoc-Version (1.x) installiert.

Wir brauchen daher folgendes Setup:
- aktuelles Ubuntu (mindestens 20.04)
- PHP 7.1
- Pandoc 2.x

Hast Du eine Idee, wie wir das mit Github Actions einfach bewerkstelligen können?

Hoffentlich brauchen wir keinen _Self Hosted Runner_ (https://github.com/marketplace/actions/setup-php-action#self-hosted-runners)?

**Comment by j3nsch** [comment]
Was uns davon abhält hier mit Ubuntu 20 zu testen, ist der Coding Style Check. Version 1.0.0 vom Laminas-Coding-Style hat immer noch die problematische Abhängigkeit,  die zu der Fehlermeldung führt. Mit der aktuellen Version des Coding Styles funktioniert es mit Ubuntu 20, aber nicht mehr mit Ubuntu 16. 

Ich denke es macht Sinn hier vorwärts zu denken und die Tests mit Ubuntu 20 und dem aktuellen Coding Style von Laminas laufen zu lassen. Der Style ändert sich damit ein wenig. Er ist strikter.

Im Augenblick kann ich das auf GitHub nicht testen, weil es Probleme mit den Actions gibt. Ich schaue mir die Änderungen im Coding Style noch einmal auf meinem System an, bevor die Entscheidung entgültig ist.

Davon abgesehen werden im Augenblick Reports generiert, die wir nicht weiterverwenden. Ein einfacher "check" anstelle von "check-full" reicht aus. Leider löst das das Problem noch nicht und wir müssen uns trotzdem zwischen alt und neu entscheiden.

Bin leider noch nicht dazu gekommen mir den Code im Detail anzuschauen.

**Comment by j3nsch** [comment]
Ich habe im "improved" Branch die Version des BibTeX Codes, die ich hatte, auf den Laminas Coding Style umgestellt und auch die PHPUnit Version aktualisiert. Da wir hier nicht von ZF1 abhängen, müssen wir uns dadurch nicht beschränken lassen. Die Änderungen solltest Du übernehmen können. 

Der neue Coding Style ist strikter, vorallem bei den Type Hints. Ich habe ihn ein wenig entschärft, damit wir PHP 7.0 kompatibel bleiben und damit wir unsere Header-Annotationen weiter verwenden können. Bie den Annotationen ist der Gedanke anscheinend, dass Git diese Informationen bereits enthält. Da muss ich später noch mal drüber nachdenken.
Ich denke die erhöhten Anforderungen machen Sinn. Der Coding Style sorgt auch für ein Alignment von Zuweisungen. Da bin ich mir noch nicht sicher, dass ich das gut finde. 

Drei Coding Style Meldungen zu Namen von Klassen und Interfaces habe ich noch offen gelassen für Dich. Ich denke die Namen sollten entsprechend geändert werden.

Wir müssen erst einmal kompatibel zu PHP 7.0 bleiben. Das spielt z.B. eine Rolle wenn Return Type Hints verwendet werden. 

Ich habe mir bei den Änderungen auch den Code angeschaut und das Feedback dazu kommt auch heute. Vielen DanK!

**Review by j3nsch — CHANGES_REQUESTED** [review]
Vielen Dank! Ich habe mir nicht jede einzelne Zeile anschaut, sondern eher auf allgemeine Konzepte geachtet. Wenn es für das Hosting funktioniert ist Funktional erst einmal alles gut.

Wie die Konfiguration jetzt funktioniert ist in Ordnung. Da wird sich später noch mal was ändern und vermutlich schaue ich mir irgendwann auch noch Vererbung wie z.B. auch bei den Coding Style Konfigurationen an.

Es wäre schön wenn die geworfenen allgemeinen \Expection durch eine konkretere Klasse im BibTeX-Paket ersetzt werden könnten, `ParserException` oder eine neue Klasse.

Das sieht alles gut aus und ich sehe im Augenblick keine Probleme, die sich später nicht leicht fixen lassen sollten, wenn sich die Anforderungen weiterentwickeln. 

Bitte die Coding Style Änderungen aus dem Branch "improved" noch übernehmen und die letzten kleinen Probleme damit fixen.

Vielen Dank!

File: src/Import/Rules/Umlauts.php [hunk 1]
```diff
@@ -33,45 +33,18 @@
...
-        );
-    }
-
+    /**
+     * Wandelt speziell angegebene Umlaute im übergebenen Wert in ihr zugehöriges Unicode-Zeichen um.
+     *
+     * @param string $value Wert, der ggf. umzuwandelnde Umlaute enthält
+     * @return string|boolean liefert den umgewandelten Wert oder false, falls keine Umwandlung von Zeichen erfolgt ist
+     */
     private function convertUmlauts($value)
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply src/Import/Rules/Umlauts.php:54]: umgesetzt

**Comment by saschaszott** [comment]
@j3nsch , wenn ich die Tests lokal ausführe, dann schlagen 10 von 107 fehl, weil die Methode `expectException` nicht bekannt ist. Hast Du einen Commit vergessen?

**Comment by saschaszott** [comment]
> @j3nsch , wenn ich die Tests lokal ausführe, dann schlagen 10 von 107 fehl, weil die Methode `expectException` nicht bekannt ist. Hast Du einen Commit vergessen?

ok, hat sich erledigt - ich habe gesehen, dass du die PHPUnit Dependency auf Version 9 hochgesetzt hast. Damit musst ich mir lokal allerdings nun zwei PHP-Versionen installieren (7.3+ damit PHPUnit 9 läuft und 7.1 damit ich die OPUS-Applikation im Browser aufrufen kann)

File: src/Import/Rules/Umlauts.php [hunk 2]
```diff
@@ -83,4 +56,37 @@ private function convertUmlauts($value)
...
     }
+
+    /**
+     * Iteriert über die im OPUS-Metadatensatz enthaltenen Felder und ersetzt speziell angegebene Umlaute durch ihr
+     * zugehöriges Unicode-Zeichen.
+     *
+     * @param array $fieldValues Wert aus BibTex-Feldern: wird an dieser Stelle nicht ausgewertet
+     * @param array $documentMetadata OPUS-Metadatensatz (Array von Feldwerten)
+     */
+    protected function setFields($fieldValues, &$documentMetadata)
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply src/Import/Rules/Umlauts.php:106]: Rekursion erhöht m.E. an dieser Stelle eher die Komplexität für den Leser des Quellcodes; ich habe die Methode jetzt auseinandergezogen

Die Blacklist kann nun konfiguriert werden.

File: README.md [hunk 3]
```diff
@@ -62,15 +62,15 @@ metadata fields. A field mapping is provided by a configuration file. At the mom
 supports configuration files in the JSON format only. It is possible to manage several
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply README.md:62]: guter Punkt; ich hatte die `README.md` in Englisch verfasst, weil sie bereits in dieser Sprache vorlag - vermutlich können wir aber auf Deutsch umstellen aufgrund des deutsprachigen Fokus von OPUS4

File: src/Import/Config/BibtexMapping.php [hunk 4]
```diff
@@ -190,26 +204,61 @@ public function resetRules()
...
-            $this->addRule($name, $class);
-            if (array_key_exists('properties', $rule)) {
-                foreach ($rule['properties'] as $propName => $propValue) {
-                    $setter = 'set' . ucfirst($propName);
-                    $class->$setter($propValue);
+            $ruleInstance = $this->getRuleInstance($rule);
+            $this->addRule($name, $ruleInstance);
+            if (array_key_exists('options', $rule)) {
+                $options = $rule['options'];
+                if ($ruleInstance instanceof ConstantValues) {
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply src/Import/Config/BibtexMapping.php:240]: man könnte auch alternativ auf das Vorhandensein der Methode `setOptions` umstellen - aber ist das "besser"? Die Spezialbehandlung für `ConstantValues` als "convenience" Regelklasse für das einfache Setzen _mehrerer_ Konstanten können wir nicht vermeiden

File: src/Import/Rules/IRule.php [hunk 5]
```diff
@@ -36,14 +36,17 @@
 interface IRule
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply src/Import/Rules/IRule.php]: erledigt

File: src/Import/Rules/Umlauts.php [hunk 6]
```diff
@@ -33,45 +33,18 @@
 
 namespace Opus\Bibtex\Import\Rules;
 
-use Opus\Bibtex\Import\Configuration\FieldMapping;
-
 /**
  * Behandlung von Umlauten, die im BibTeX-File nicht korrekt angegeben wurden (siehe OPUSVIER-4216).
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply src/Import/Rules/Umlauts.php:46]: Ich habe dafür den Verweis auf die Testdatei `specialchars-invalid.bib` angegeben. Mehr kann ich dazu nicht sagen, da die Umlaute eine Besonderheit im BibTeX-File von Thorsten sind. Wie / Warum diese Umlaute entstehen, ist mir unklar. Die Referenz auf das Jira-Ticket würde ich ungern entfernen.

File: test/Import/Rules/LanguageTest.php [hunk 7]
```diff
@@ -24,31 +24,28 @@
...
- * @package     Opus\Bibtex\Import\Rules
+ * @category    Tests
+ * @package     OpusTest\Bibtex\Import\Rules
  * @author      Sascha Szott <opus-repository@saschaszott.de>
  * @copyright   Copyright (c) 2021, OPUS 4 development team
  * @license     http://www.gnu.org/licenses/gpl.html General Public License
  */
 
-namespace Opus\Bibtex\Import\Rules;
+namespace OpusTest\Bibtex\Import\Rules;
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply test/Import/Rules/LanguageTest.php:36]: ja, das war mir auch aufgefallen - die Ursache habe ich bislang nicht weiter erforscht

File: src/Import/Rules/Pages.php [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            $pageNumber = $this->getPageNumber($documentMetadata);
+            if (! is_null($pageNumber)) {
+                $documentMetadata['PageNumber'] = $pageNumber;
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private function getPageFirst($value)
```
[the remark this answers is not in the document]
Comment (saschaszott) [reply src/Import/Rules/Pages.php]: umgesetzt

**Comment by saschaszott** [comment]
> Es wäre schön wenn die geworfenen allgemeinen \Expection durch eine konkretere Klasse im BibTeX-Paket ersetzt werden könnten, `ParserException` oder eine neue Klasse.

es wurde eine neue Exception-Klasse eingeführt

**Comment by saschaszott** [comment]
@j3nsch , ich habe deine Kommentare / Verbesserungsvorschläge eingearbeitet - bitte schau nochmal kurz drüber und setze die einzelnen Punkte bitte auf erledigt

Ich würde nun noch die 3 offenen TODOs von oben angehen.

**Comment by saschaszott** [comment]
@j3nsch , mir ist aufgefallen, dass in den Test-Klassen an einigen Stellen nun _Type Hints_ stehen, z.B. in 

```php
class TitleTest extends TestCase
{
    public function dataProvider(): array
```

Das war so beabsichtigt?

**Comment by j3nsch** [comment]
> > @j3nsch , wenn ich die Tests lokal ausführe, dann schlagen 10 von 107 fehl, weil die Methode `expectException` nicht bekannt ist. Hast Du einen Commit vergessen?
> 
> ok, hat sich erledigt - ich habe gesehen, dass du die PHPUnit Dependency auf Version 9 hochgesetzt hast. Damit musst ich mir lokal allerdings nun zwei PHP-Versionen installieren (7.3+ damit PHPUnit 9 läuft und 7.1 damit ich die OPUS-Applikation im Browser aufrufen kann)

Ja, ist blöd momentan. Ich arbeite auf zwei Systemen, Ubuntu 16 und Ubuntu 20. Ich habe versucht auf PHPUnit 7.5 runter zu gehen. Dort gibt es bereits `expectExpectation` und es läuft noch mit PHP 7.1. Dann klappt aber der Build mit Ubuntu 20 auf GitHub nicht mehr, da dort anscheinend schon PHP 8 zum Einsatz kommt. 

Im "improved" Branch habe ich jetzt die Tests mit Ubuntu 20, PHP 7.4.18 und PHPUnit 7 laufen. Leider habe ich nicht herausgefunden, ob und wie ich das mit `update-alternatives`, also mit Bordmitteln, machen kann. Jetzt wird eine 3rd Party Action "Setup PHP" eingesetzt. Ich denke das ist in Ordnung.

File: src/Import/Rules/Umlauts.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Import/Rules/Umlauts.php:46]: Danke, ist ja auch gut, dass die Ticket-Referenz drin ist, aber eben nur für uns. Ich denke die Beispieldatei sollte reichen.

**Comment by j3nsch** [comment]
Jetzt habe ich auch die aktuellen Änderungen in den "improved" Branch übernommen. Dort läuft jetzt alles auch mit PHPUnit 7.5 durch. Damit solltest Du eigentlich wieder nur mit PHP 7.1 entwickeln können.

**Comment by j3nsch** [comment]
> @j3nsch , mir ist aufgefallen, dass in den Test-Klassen an einigen Stellen nun _Type Hints_ stehen, z.B. in
> 
> ```
> class TitleTest extends TestCase
> {
>     public function dataProvider(): array
> ```
> 
> Das war so beabsichtigt?

Ja, mit dem neuen Coding Style hat man die Wahl zwischen Annotation (`@return`) ODER Type Hint. Wenn beides angegeben wird, gibt es wohl auch Beschwerden. Ich sehe erst einmal kein Problem mit dem Type Hint, gerade bei den Tests. Im normalen Code sind mit Annotationen lieber. Was man beachten muss, ist, dass zwar der Type Hint Array für Rückgabewerte schon in PHP 7.0 verfügbar war, andere Typen aber durchaus erst später dazu gekommen sind.

**Comment by saschaszott** [comment]
> Jetzt habe ich auch die aktuellen Änderungen in den "improved" Branch übernommen. Dort läuft jetzt alles auch mit PHPUnit 7.5 durch. Damit solltest Du eigentlich wieder nur mit PHP 7.1 entwickeln können.

Vielen Dank. Ich habe deine Änderungen aus dem Branch `improved` übernommen. Mit PHPUnit 7.5 (Github Action nutzt 7.5.20) laufen die Tests erfolgreich durch.

File: src/Import/Rules/Umlauts.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Import/Rules/Umlauts.php:106]: Danke.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (j3nsch) [reply README.md:62]: Siehst Du und ich neige dazu eher generell in Richtung Englisch zu gehen. Wir lassen es jetzt erst mal so wie es ist.

File: src/Import/Config/BibtexMapping.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (j3nsch) [reply src/Import/Config/BibtexMapping.php:240]: Auf `setOptions` zu prüfen, würde auch funktionieren, ist aber unschön. Damit die Zusammenhänge klar sind, müsste die Funktion Teil des Interfaces sein. Die Verantwortung für das Setzen würde sich dann in die Regelklassen verschieben. Wir würden eine Basisklasse mit dem Standardmechanismus benötigen, was nicht schlimm ist, aber vielleicht auch nicht notwendig. 

Wie wäre es denn, wenn wir Options so lassen wie es ist und die Klasse `ConstantValues` einen Setter, für sagen wir mal `fields` oder `values` (or whatever), bekommt. 

```
  "options" : {
    "fields" : {
       "language": "eng" 
       "belongsToBibliography": "true"
    }    
  }
```

Das sollte dann eigentlich als Array ankommen und an den Setter übergeben werden. Das JSON ist nicht ganz so schön, aber wir brauchen keine Sonderbehandlung.

Die Klasse wird im Augenblick aber gar nicht verwendet. Das ist also eher theoretisch. Es kann auch einfach so bleiben wie es ist und wir schauen mal wie sich die Nutzung entwickelt. Wie gesagt, ich würde sowieso später gerne noch Vererbung einbauen.

**Comment by j3nsch** [comment]
Ich glaube das sieht gut aus.

**Review by j3nsch — APPROVED** [review]
Ich übernehme das jetzt einfach mal. Vielen Dank!

File: src/Import/Config/BibtexMapping.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (saschaszott) [reply src/Import/Config/BibtexMapping.php:240]: ok, dann belassen wir es erstmal so - spätere Verbesserungen sind ja jederzeit umsetzbar, sofern Bedarf besteht
